File: 双指针/java/680.验证回文字符串-ⅱ.java
/*
 * @lc app=leetcode.cn id=680 lang=java
 *
 * [680] 验证回文字符串 Ⅱ
 */

// @lc code=start
class Solution {
    public boolean validPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                break;
            }
            left++;
            right--;
        }
        return partValidPalindrome(s, left + 1, right) || partValidPalindrome(s, left, right - 1);
    }

    public boolean partValidPalindrome(String s, int left, int right) {
        // delete right char
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
}
// @lc code=end

File: 双指针/python/680.验证回文字符串-ⅱ.py
#
# @lc app=leetcode.cn id=680 lang=python3
#
# [680] 验证回文字符串 Ⅱ
#

# @lc code=start
class Solution:
    def partValidPalindrome(self, s: str, left: int, right: int) -> bool:

            while left < right:
                if s[left] != s[right]:
                    return False
                left = left + 1
                right = right -1
            
            return True

    def validPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        while left < right:
            if s[left] != s[right]:
                break
            left = left + 1
            right = right - 1
        return self.partValidPalindrome(s, left + 1, right) or self.partValidPalindrome(s, left, right - 1)

# @lc code=end

